MicroPython examples

stepansnigirev opened the pull request:
Micropython examples for most of the object types.
Related discussion: https://forum.littlevgl.com/t/examples-for-lvgl-in-micropython/1190

Not implemented examples: 
- **arc - 2nd example** - uses `lv.task_t`, I was not able to make it work :(
- **calendar** - `highlighted_dates` method doesn't accept a list
- **img** - not sure how to provide image data yet, probably makes sense to use png loader or a hardcoded bytearray
- **imgbtn** - same as above
- **label - 3rd example** - similar to `arc` - also uses `lv.task_t`
- **tileview** - requires passing an array of `lv_point_t`, list doesn't work. Very similar to `calendar` problem.

Method `get_selected_str()` in `ddlist` and `roller` works but pretty ugly - it requires to pass a string as an argument and changes part of it in place. 
Therefore in the examples I use the following callback:
```py
def event_handler(obj, event):
    if event == lv.EVENT.VALUE_CHANGED:
        idx = obj.get_selected()
        option = obj.get_options().split("\n")[idx]
        print("Option: %s" % option)
```

Also I don't know how to get data from `lv.event_get_data()` (`ta` example)

embeddedt commented:
@amirgon Can you review this?

amirgon commented:
@stepansnigirev Nice work!

Since the simulator is currently based on older code, I suggest testing these scripts on the unix (linux) simulator on [`dev-6.1` branch of `lv_micropython`](https://github.com/littlevgl/lv_micropython/tree/dev-6.1).  
`dev-6.1` will be merge very soon with master so it makes sense working on it and building the simulator from it.

@stepansnigirev were you able to build and run it on your side?

> * **arc - 2nd example** - uses `lv.task_t`, I was not able to make it work :(
> * **label - 3rd example** - similar to `arc` - also uses `lv.task_t`

On the latest version, Tasks work.
I've implemented **arc - 2nd example** here: https://gist.github.com/amirgon/ececcec778fb089a6ddc7dbf728d2ec2
Please have a look, it work well on `dev-6.1`. (You can take it as is if you want)

It shows:

- How to use lv_task. Actually there are 2 ways to do it, with `lv_task_create` or with `lv_task_create_basic`.
- How to model the code in a more Object Oriented way. Since it's python, why write it like C?  
In this example I created a new class `loader_arc` that inherits from `lv.arc` and extends it.

> * **calendar** - `highlighted_dates` method doesn't accept a list

As discussed, we would need to change `highlighted_dates` prototype on lvgl sources, and perhaps other functions that receive a pointer instead of an array.  
I will create a PR for this.

> * **img** - not sure how to provide image data yet, probably makes sense to use png loader or a hardcoded bytearray
> * **mgbtn** - same as above

You can load files from the network the same way you can import from the network.  
Just need to provide the full URL as the filename and make sure the web server support [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) (if you stick to github rawcontent you'll be fine, it supports CORS)

[This example](https://littlevgl.github.io/lv_micropython/ports/javascript/lvgl_editor.html?script_direct=3b62221f160028a418f30a4bf17f9ca729ff1536) shows how to load image from the web (it also shows how to display raw image and PNG image).  
It takes some time to load on the web browser, be patient.

The relevant line is
```python
with open('https://raw.githubusercontent.com/littlevgl/lv_binding_micropython/master/examples/png_decoder_test.png','rb') as f:
  png_data = f.read()
```
> * **tileview** - requires passing an array of `lv_point_t`, list doesn't work. Very similar to `calendar` problem.

Should work on `dev-6.1`

> Method `get_selected_str()` in `ddlist` and `roller` works but pretty ugly - it requires to pass a string as an argument and changes part of it in place.

In principle strings are supposed to be immutable.  
But in this case it's so much easier to initialize a string and pass it. This is how C expects it to work, and it works here as well.  
My opinion is that it's better to pass a string than create workarounds like you suggested here.  
We can consider passing a bytearray and convert it to string, if this is important.

> Also I don't know how to get data from `lv.event_get_data()` (`ta` example)

In the case of `ta` example, you need to cast a pointer to string.
It's a bit tricky in Micropython, but here is an example of how this can be done:

```python
import display_driver
import lvgl as lv

def cb(ta, event):
    if event == lv.EVENT.INSERT:
        ptr = lv.C_Pointer()
        ptr.ptr_val = lv.event_get_data()
        print(ptr.str_val)

ta = lv.ta(lv.scr_act())
ta.set_event_cb(cb)
```

The `C_Pointer` struct (union actually) is used to cast a pointer to other types, such as string.

stepansnigirev commented:
I checked on dev-6.1 branch and added a few missing examples.
Calendar, Tileview and Tasks work - that's awesome!
I also added a png example in `lv_ex_img`, but because of header parsing and decoding, this example becomes not so minimal...

For `lv_arc` I took an example from @amirgon , I also refactored the `mbox` modal example with a class. 

One more issue I found:
`calendar.get_pressed_date()` doesn't return `None` if the header is clicked (on C side it returns NULL but it is not converted to None)

> I also noticed lv_mbox_add_btns, I think your modal message box demo doesn't work because of this.
> Did you find more examples?

Actually `mbox` demo works fine: [demo](https://littlevgl.github.io/sim/micropython/ports/javascript/lvgl_editor.html?script=https://gist.githubusercontent.com/stepansnigirev/bd0f2c4c2b95b530f28ec98a441311c2/raw)

kisvegabor commented:
@stepansnigirev 
You made awesome progress. 
Congratulations, and thank you! :slightly_smiling_face: 

I'll merge it if @amirgon approves it.

amirgon commented on src/lv_ex_mbox/lv_ex_mbox_2.py:
```diff
@@ -0,0 +1,86 @@
+
+        # Create the message box as a child of the modal background
+        mbox = Modal(lv.scr_act())
+        mbox.add_btns(btns2)
```
Is `add_btns` working for you on `dev-6.1`? I'm getting: `SyntaxError: Cannot convert 'list' to pointer!`

`add_btns` [prototype](https://github.com/littlevgl/lvgl/blob/b4b6a472fb54c4f11b848fd3ee3910b65d6fee79/src/lv_objx/lv_mbox.h#L97) is:
```c
void lv_mbox_add_btns(lv_obj_t * mbox, const char ** btn_mapaction);
```

@kisvegabor - probably this too needs to be changed to an array of char* instead of a pointer to char*, for the binding to recognize this as an array and not a pointer.

stepansnigirev replied:
Yes, it works for me both on unixport and in the javascript simulator... Using `dev-6.1`.
[online simulator](https://littlevgl.github.io/sim/micropython/ports/javascript/lvgl_editor.html?script_direct=786e686a3a4ce922fe9aaaa3ea9a4b13315a7cfa)

amirgon replied:
> Yes, it works for me both on unixport and in the javascript simulator... Using `dev-6.1`.
> [online simulator](https://littlevgl.github.io/sim/micropython/ports/javascript/lvgl_editor.html?script_direct=786e686a3a4ce922fe9aaaa3ea9a4b13315a7cfa)

The online simulator on https://littlevgl.github.io/sim/ is not up to date, AFAIK.  
But if you are saying this is working for you on the unix simulator with up-to-date dev-6.1... Maybe some local problem on my side. I'll recheck.

amirgon replied:
Checked again - same problem. I think we are not aligned to the same lvgl vesion. Did you update submodules?  
Anyway, I've submitted a PR for fixing this on lvgl: https://github.com/littlevgl/lvgl/pull/1279

pmp-p commented on src/lv_ex_img/lv_ex_img_1.py:
```diff
@@ -0,0 +1,94 @@
+
+# Create a screen with a draggable image
+
+with open('cogwheel.png','rb') as f:
```
@amirgon you should be able do that from the js file api every **python** "open" leads there https://github.com/littlevgl/lv_micropython/blob/lvgl_javascript/ports/javascript/docs/wasm_file_api.js#L51 if you ad `:` as prefix to filename,  but it should not be made via sys.path. modification since only import cares about that.  Maybe look into micropython VFS commands and emulate them for the simulator ?

stepansnigirev replied:
I think we could ask javascript to fetch files from the script path.
If simulator has a script in url: `script=path/to/script/name.py` and we try to open `cogwheel.png` simulator should try to fetch `path/to/script/cogwheel.png`.

pmp-p replied:
indeed with a filter - especially to avoid doing that on .py files - restricted to png/jpg it could work without changing C code on js side.

embeddedt replied:
Yes, opening files synchronously on the web is not the best, but for our examples, it is simpler and should satisfy the majority of users' needs.

stepansnigirev commented on src/lv_ex_img/lv_ex_img_1.py:
```diff
@@ -0,0 +1,94 @@
+
+# Parse PNG file header
+# Taken from https://github.com/shibukawa/imagesize_py/blob/ffef30c1a4715c5acf90e8945ceb77f4a2ed2d45/imagesize.py#L63-L85
+def get_png_info(decoder, src, header):
```
I will be happy to remove these functions from the example. They make it too complex.
How should we call this module? `imagetools`?
I could do something like this: 
```py
from imagetools import get_png_info, open_png
# Register new image decoder
decoder = lv.img.decoder_create()
decoder.info_cb = get_png_info
decoder.open_cb = open_png
# ... and so on from there
```

amirgon replied:
> How should we call this module? `imagetools`?
> I could do something like this:

Yes, that sounds good.
Could you send a PR to [lv_micropython_binding dev-6.1](https://github.com/littlevgl/lv_binding_micropython/tree/dev-6.1)?

stepansnigirev replied:
Here is the pull request: https://github.com/littlevgl/lv_binding_micropython/pull/52
Moved decoding functions from the example file in https://github.com/littlevgl/lv_examples/pull/40/commits/5963b319e82a57f5844144d3cffdc406efac3d7e

stepansnigirev commented on src/lv_ex_calendar/lv_ex_calendar_1.py:
```diff
@@ -0,0 +1,35 @@
+calendar.set_showed_date(today)
+
+# Highlight some days
+highlihted_days = [lv.calendar_date_t() for i in range(3)]
```
That's way better! Fixed.

amirgon commented:
> > `calendar.get_pressed_date()` doesn't return `None` if the header is clicked (on C side it returns NULL but it is not converted to None)
> 
> It should be quite simple to fix.
> I've opened a new issue: [littlevgl/lv_binding_micropython#50](https://github.com/littlevgl/lv_binding_micropython/issues/50)

Fixed.

kisvegabor commented:
> Anyway, I've submitted a PR for fixing this on lvgl: littlevgl/lvgl#1279

Merged!

amirgon commented:
The micropython examples themselves look good, I think they can be merged.
Thanks @stepansnigirev!

Open issues are related to the online simulator:
- The `rst` files should also contain the links to the online simulator for each example.
Currently they only contain the example code.
- In case of loading an image, the simulator needs to load it relative to some URL, as [discussed](https://github.com/littlevgl/lv_examples/pull/40#discussion_r349356269).
- The simulator itself would need to be re-generated with the latest code on `lv_micropython`.

kisvegabor commented:
>  I think they can be merged.

I merged it so it will be part of the documentation of `v6.0` (released today).

@stepansnigirev Thank you very much for adding these examples! :clap: :clap: :clap:
